### PR TITLE
SCUMM: LOOM/GUI: Implement a Drafts Inventory

### PR DESCRIPTION
**Pull request by AndywinXp**
## Hey you! _Yes, you!_

Have you ever loaded your old LOOM savegames from when you were just a baby, and promptly bursted into tears not because of nostalgia but because you **_couldn't remember the dang drafts of that playthrough_**?

Well, cry no more!

Introducing the:

## ...what?

...more seriously though, this PR adds an enhancement to LOOM which allows the user to view which drafts have currently been unlocked on the current playthrough. This is basically the "drafts" command of the debugger, but on steroids. 🙂

This idea is not mine, but it's actually Aric Wilmunders'! I heard it on a podcast interview he did a few days ago and thought that it did seem feasible...

## How do I use it?

Just activate the enhancements checkbox and the original GUI for the game, go in-game, press CTRL-D, and it will come up!

##  Which versions are supported?

All of them, and for every language, except:
* PC-Engine: out of the scope of this PR, it doesn't even properly support GUI elements for now;
* Mac: I would like to do a full-on MacGui layout for this, but I would like to wait for eriktorbjorn's new menu bar changes to be in tree before I attempt this; so this is currently out of scope.

## How does it look?
scummvm-loom-ega-00001
scummvm-loom-ega-fr-00001
scummvm-loom-ega-de-00000
scummvm-loom-ega-he-00001
scummvm-loom-ega-es-00000
scummvm-loom-fm-ja-00000
scummvm-loom-fm-00001
scummvm-loom-steam-win-00000

Like this! Drafts can be:
* Locked
* Unlocked but never used
* Unlocked and used at least once

Let me know what you think!

### CODE REVIEW

**Comment by bluegr** [comment]
Wow, great work! Kudos, well done! :)

**Comment by eriktorbjorn** [comment]
"Mac: I would like to do a full-on MacGui layout for this"

A desktop accessory? ;-)

image

**Comment by AndywinXp** [comment]
> "Mac: I would like to do a full-on MacGui layout for this"
> 
> A desktop accessory? ;-)
> 
> image

...you just read my mind, Sir 😄 that's *exactly* what I was thinking about!

**Comment by eriktorbjorn** [comment]
One thing that I didn't bother about when I wrote the "drafts" debugger command.

The physical "Book of Patterns" includes patterns that aren't in the game as red herrings, so technically listing only the ones that are actually in the game is a bit of a spoiler. Perhaps it should only list the ones you have discovered so far, and keep the rest hidden?

The ones in the book are:

- Opening
- Dyeing
- Summoning
- Tongues
- Straw Into Gold
- Temblor
- Rending
- Healing
- Night Vision
- Shrinkage
- Desire
- Waterproofing
- Reflection
- Terror
- Folding
- Invisibility
- Twisting
- Confusion
- Shaping
- Warmth
- Aphrodesia
- Sleep
- Silence
- Emptying
- Unmaking
- Extinguishing
- Sharpening
- Blessing
- Transcendence

So including the red herrings as well would just make a clutter of things.

**Comment by BLooperZ** [comment]
maybe build the list dynamically? (e.g
all lines are blank until heard, then each line is added as drafts are discovered)

**Comment by AndywinXp** [comment]
I though about that, but unfortunately the very first and the very last drafts are immediately given as learned to the player, so this means that I would have something like this:
image

which looks kind of weird on a design perspective...

**Comment by eriktorbjorn** [comment]
Couldn't you just draw it like this? Or is there a reason the drafts have to be on fixed positions?
image

**Comment by AndywinXp** [comment]
> Couldn't you just draw it like this? Or is there a reason the drafts have to be on fixed positions?

Unfortunately this would require a major rewrite of the layout code...

Also, I think I want to make it more modern and appealing, design-wise; allow me to give an example:
image
This looks like a list of attributes and skills, kind of like a Dungeon and Dragons character sheet. Which is fine.


On the other hand:
scummvm-loom-ega-00002
This makes it look as if you have unlockable power-ups, which also brings a psychological element with it: it makes the player push forward and explore everything until they have unlocked everything.

Of course you're bound to unlock all drafts in this game, but given that this PR aims at making the game a little bit more accessible, I also find myself wanting to push towards these small modern details.

That's only my own take, though 🙂

**Comment by bluegr** [comment]
I really like the last change (i.e. hide the currently unknown drafts titles), and I believe that it addresses @eriktorbjorn's point about the red herring drafts.

I also like the "unlockable powerups" style of the drafts dialog. IMHO, it gives the player an overview of the drafts, with a sense of progress and previously unlocked drafts, without giving away which ones are really used in the game.

Overall, I believe this is a very nice and useful enhancement screen!

File: engines/scumm/gfx_gui.cpp [hunk 1]
```diff
@@ -1666,31 +1652,38 @@ static const char *loomDraftsNames[6][17] = {
 	// HEBREW
 	{
 		"\x9a\x85\x90\x89\x82\x90\x8e",
```
Comment (OmerMor) [inline engines/scumm/gfx_gui.cpp:1725]: FWIW, the Hebrew translation to "Unknown" is "לא ידוע" and the string for it should be:
`":\x84\x85\x83\x89 \x80\x8c"`

Comment (AndywinXp) [reply engines/scumm/gfx_gui.cpp:1725]: Thank you, I think I already have it as the last element of the array (I commited it yesterday), and this is how it looks:
image

Comment (OmerMor) [reply engines/scumm/gfx_gui.cpp:1725]: > Thank you, I think I already have it as the last element of the array (I commited it yesterday), and this is how it looks: image

Looks great!

**Comment by sev-** [comment]
Indeed, spoler-free version, e.g. showing "Unknown: ???" for unseen drafts is the way to go :)

**Comment by sev-** [comment]
I will provide draft names for Russian fan-translation from the RuSCUMM project.

**Comment by bluegr** [comment]
So, this looks OK to be merged now, I believe

Adding extra translations can be done in-tree.
Any objections in merging it?

**Comment by sev-** [comment]
No objections, merging. Excellent work! I will put Russian in-tree, I just got the strings, but I need to adapt to the encoding.

**Comment by AndywinXp** [comment]
Thank you! 😄

**Pull request merged**